### PR TITLE
CLICKHOUSE-3211 add TOP m, and OFFSET k

### PR DESCRIPTION
**Pull request by VadimPlh**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

**Comment by zhang2014** [comment]
What does this do? It looks like it's the same as the `limit`

**Comment by VadimPE** [comment]
@zhang2014 You are right) It seems like a `LIMIT`

**Comment by zhang2014** [comment]
So why do we need it?

**Comment by VadimPE** [comment]
@zhang2014 
image

**Comment by VadimPE** [comment]
image

**Comment by VadimPE** [comment]
@zhang2014
For TOP ... WITH TIES

**Comment by derekperkins** [comment]
Does this solve this issue in https://github.com/yandex/ClickHouse/issues/2198?

**Comment by alexey-milovidov** [comment]
> Does this solve this issue in #2198?

No. The task #2198 is still unassigned.

**Comment by alexey-milovidov** [comment]
> So why do we need it? I'm just curious : )

LIMIT OFFSET is required for compatibility.
This syntax exist in PostgreSQL and was standardized in SQL 2008.
Some tools will generate queries using this syntax (instead of `LIMIT n, m`).
It also was added in MySQL with the same motivation.

And `SELECT TOP` is a bonus (compatibility with MS SQL + allows the implementation `WITH TIES`. We could add `WITH TIES` also after `LIMIT` but it's more known in MS SQL `TOP` syntax).

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
We should throw an exception if offset is specified in both forms `LIMIT 10, 20 OFFSET 30` or if both `LIMIT` and `TOP` are specified.

Test should check that exceptions are thrown.

File: dbms/tests/queries/0_stateless/00687_top_and_offset.sql [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+INSERT INTO test.test VALUES (2);
+INSERT INTO test.test VALUES (3);
+INSERT INTO test.test VALUES (4);
+INSERT INTO test.test VALUES (5);
+INSERT INTO test.test VALUES (6);
+INSERT INTO test.test VALUES (7);
+INSERT INTO test.test VALUES (8);
+INSERT INTO test.test VALUES (9);
+
+SELECT TOP 2 * FROM test.test;
```
Comment (alexey-milovidov) [inline dbms/tests/queries/0_stateless/00687_top_and_offset.sql]: The test is non-deterministic due to lack of ORDER BY.

File: dbms/tests/queries/0_stateless/00687_top_and_offset.sql [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+SELECT TOP (2) * FROM test.test;
+SELECT * FROM test.test LIMIT 2 OFFSET 2;
+SELECT TOP 2 * FROM test.test LIMIT 2; -- { clientError 406 }
```
Comment (alexey-milovidov) [inline dbms/tests/queries/0_stateless/00687_top_and_offset.sql]: Negative test for `LIMIT n, m OFFSET k` is pending.

**Pull request merged**